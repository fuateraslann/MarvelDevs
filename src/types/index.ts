export type TImageExtensions = 'jpg' | 'png' | 'svg'
export type TCharacter = {
  id: number
  name: string
  description: string
  thumbnail: {
    path: string
    extension: TImageExtensions
  }
}
export type TComics = {
  id: number
  title: string
  thumbnail: {
    path: string
    extension: TImageExtensions
  }
}
