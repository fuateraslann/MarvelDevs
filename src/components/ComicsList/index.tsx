import { IconButton, ImageList, ImageListItemBar, Tooltip } from '@mui/material'
import ImageListItem from '@mui/material/ImageListItem/ImageListItem'
import InfoIcon from '@mui/icons-material/Info'
import { TComics } from 'types'
import { isEmpty } from 'lodash'

const ComicsList = ({
  isLoading,
  comics,
}: {
  isLoading: boolean
  comics: TComics[]
}) => {
  if (isLoading) return <div>Loading...</div>
  if (isEmpty(comics)) return <div>No comics found</div>
  return (
    <ImageList
      sx={{
        gridAutoFlow: 'column',
        gridTemplateColumns: 'repeat(auto-fill,minmax(160px,1fr)) !important',
        gridAutoColumns: 'minmax(160px, 1fr)',
      }}
      data-testid="comics-list"
    >
      {comics?.map((e) => {
        return (
          <ImageListItem key={e.id}>
            <img src={e.thumbnail.path + '.' + e.thumbnail.extension} />
            <ImageListItemBar
              data-testid="image-list-item-bar"
              title={e.title}
              actionIcon={
                <Tooltip title={e.title}>
                  <IconButton
                    sx={{ color: 'rgba(255, 255, 255, 0.54)' }}
                    aria-label={`info about ${e.title}`}
                  >
                    <InfoIcon />
                  </IconButton>
                </Tooltip>
              }
            />
          </ImageListItem>
        )
      })}
    </ImageList>
  )
}

export default ComicsList
