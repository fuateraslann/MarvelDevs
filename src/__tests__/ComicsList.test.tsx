import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import ComicsList from 'components/ComicsList'
import { TImageExtensions } from 'types'

const mockComics = [
  {
    id: 1,
    title: 'Comic 1',
    thumbnail: {
      path: 'test/path1',
      extension: 'jpg' as TImageExtensions,
    },
  },
  {
    id: 2,
    title: 'Comic 2',
    thumbnail: {
      path: 'test/path2',
      extension: 'jpg' as TImageExtensions,
    },
  },
]

describe('ComicsList', () => {
  test('renders ComicsList correctly with comics', () => {
    const isLoading = false

    render(<ComicsList isLoading={isLoading} comics={mockComics} />)

    expect(screen.getByTestId('comics-list')).toBeInTheDocument()
    expect(screen.getAllByTestId('image-list-item-bar')).toHaveLength(2)
    expect(screen.queryByText('Loading...')).not.toBeInTheDocument()
    expect(screen.queryByText('No comics found')).not.toBeInTheDocument()
  })

  test('renders ComicsList correctly when loading', () => {
    const isLoading = true
    render(<ComicsList isLoading={isLoading} comics={[]} />)
    expect(screen.getByText('Loading...')).toBeInTheDocument()
    expect(screen.queryByTestId('comics-list')).not.toBeInTheDocument()
    expect(screen.queryByTestId('image-list-item-bar')).not.toBeInTheDocument()
    expect(screen.queryByText('No comics found')).not.toBeInTheDocument()
  })

  test('renders ComicsList correctly with no comics', () => {
    const isLoading = false
    render(<ComicsList isLoading={isLoading} comics={[]} />)
    expect(screen.getByText('No comics found')).toBeInTheDocument()
    expect(screen.queryByTestId('comics-list')).not.toBeInTheDocument()
    expect(screen.queryByTestId('image-list-item-bar')).not.toBeInTheDocument()
    expect(screen.queryByText('Loading...')).not.toBeInTheDocument()
  })

  test('matches snapshot', () => {
    const isLoading = false

    const { asFragment } = render(
      <ComicsList isLoading={isLoading} comics={mockComics} />
    )
    expect(asFragment()).toMatchSnapshot() // Snapshot test
  })
})
