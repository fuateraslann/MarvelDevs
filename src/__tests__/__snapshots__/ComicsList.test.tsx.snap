// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComicsList matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="MuiImageList-root MuiImageList-standard css-610zjv-MuiImageList-root"
    data-testid="comics-list"
    style="grid-template-columns: repeat(2, 1fr); gap: 4px;"
  >
    <li
      class="MuiImageListItem-root css-l086p4-MuiImageListItem-root"
      style="height: auto; grid-column-end: span 1; grid-row-end: span 1;"
    >
      <img
        class="MuiImageListItem-img"
        src="test/path1.jpg"
      />
      <div
        class="MuiImageListItemBar-root MuiImageListItemBar-positionBottom css-ztrqiq-MuiImageListItemBar-root"
        data-testid="image-list-item-bar"
      >
        <div
          class="MuiImageListItemBar-titleWrap MuiImageListItemBar-titleWrapBottom MuiImageListItemBar-titleWrapActionPosRight css-14q7c0c-MuiImageListItemBar-titleWrap"
        >
          <div
            class="MuiImageListItemBar-title css-dasnyc-MuiImageListItemBar-title"
          >
            Comic 1
          </div>
        </div>
        <div
          class="MuiImageListItemBar-actionIcon MuiImageListItemBar-actionIconActionPosRight css-mjbero-MuiImageListItemBar-actionIcon"
        >
          <button
            aria-label="info about Comic 1"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1lujp9z-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="InfoIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </li>
    <li
      class="MuiImageListItem-root css-l086p4-MuiImageListItem-root"
      style="height: auto; grid-column-end: span 1; grid-row-end: span 1;"
    >
      <img
        class="MuiImageListItem-img"
        src="test/path2.jpg"
      />
      <div
        class="MuiImageListItemBar-root MuiImageListItemBar-positionBottom css-ztrqiq-MuiImageListItemBar-root"
        data-testid="image-list-item-bar"
      >
        <div
          class="MuiImageListItemBar-titleWrap MuiImageListItemBar-titleWrapBottom MuiImageListItemBar-titleWrapActionPosRight css-14q7c0c-MuiImageListItemBar-titleWrap"
        >
          <div
            class="MuiImageListItemBar-title css-dasnyc-MuiImageListItemBar-title"
          >
            Comic 2
          </div>
        </div>
        <div
          class="MuiImageListItemBar-actionIcon MuiImageListItemBar-actionIconActionPosRight css-mjbero-MuiImageListItemBar-actionIcon"
        >
          <button
            aria-label="info about Comic 2"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1lujp9z-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="InfoIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
